keps/0031-namespace-management.md

```diff
@@ -0,0 +1,152 @@
+
+```
+# operator.yaml
+name: "kubeflow"
+operatorVersion: "0.2.0"
+kudoVersion: "0.10.1"
+kubernetesVersion: 1.15.0
+appVersion: 1.0.0
+namespaceManifest: templates/namespace.yaml
+```
+
+The follow rules apply:
+
+1. Missing `namespaceManifest` means that there is no metadata (creation still applies if warranted)
+1. Missing `name: {{ .Namespace }}` is a failure (and will be validated by package verifier)
+1. No support for static namespace, it must be templated or the manifest should not exist
+1. If no `namespaceManifest` and no `--namespace` the operator installs to "default"
+1. If `--namespace` provided and namespace doesn't exist, KUDO creates it
+1. If KUDO creates namespace, if there is a `namespaceManifest` it uses it to create namespace otherwise does a simple namespace creation.
+1. If KUDO has `namespaceManifest` the check on the namespace is a deepcopy (mean it will ensure that the labels and annotations exists and match) if not it overwrites
+1. if a namespace is created KUDO waits until the namespace is created to move forward in a plan.
+1. if a namespace is needed to be created, it is detected early in the process prior to the deployment plan or as a first step.
+
+#### Alternative Ideas
+
+1. Do we allow for a default namespace other than "default"
+1. We could have another flag to enforce creation, meaning `--namespace` by itself will use that namespace but does NOT create and in combination with `--force` or `--ns-ensure` will create namespace.
+1. namespace comparison could be simple check.  It is unclear when an overwrite should occur or if this should be an error.
+
+### Proposal: Multi-Namespace (static)
+
+There is a desire to support multiple namespaces.  In this model, the expectations are that the 2nd tier namespaces (not the "primary" operator) will have statically defined namespaces which should be honored.   A significant challenge to this is objects with No namespace will land whereever KUDO is configured to install the operator, but other objects will have no control and will be static.   Part of the justification of this is kubeflow, where there are 30+ operators, many which are supportive in nature that the operator developer would like to section off. 
+
+This propose is for KUDO to deploy objects to KUDO managed {{.Namespace}} if not specified and to honor `namespace:` for objects that have them.  There are challenges to validation as we won't know what is accidently missing or accidently specified.  There are challenges to it being static which can be overcome through the use of Params.  It seems if Params are used, that they are  immutable after installation.
+
+
+### Proposal: Dependencies and Creating Namespaces
+
+Regarding dependencies and creating namespaces, it is expected that the use of `--namespace` applies to the parent AND all dependent operators AND take the namespace is created prior to applying any dependent steps.
```

Comment:
Is it expected? If I were to develop a meta-operator like kubeflow, I'd like to be able to specify that specific dependent operators are installed in different namespaces.

Reply:
Maybe we should mention that dependencies operators only support one namespace for now and that we will revisit this decision in the future when we have clear requirements.

Reply:
this was written with dependency KEP in mind and for ease of dev now as @zen-dog stated.  I will add a few words.